### DISCUSSION TITLE
Calibrated URDFs

### DISCUSSION THREAD

**Discussion by gavanderhoorn** [Yaskawa-Global/motoros2#210] [Ideas]
This is a continuation of ros-industrial/motoman#459 and #206 (specifically: https://github.com/Yaskawa-Global/motoros2/issues/206#issuecomment-1921408652).

This OP will be be expanded at a later time.

**Comment by gavanderhoorn** [comment]
@iydv wrote (in https://github.com/Yaskawa-Global/motoros2/issues/206#issuecomment-1924241769):
> According to this comment [ros-industrial/motoman#459 (comment)](https://github.com/ros-industrial/motoman/discussions/459#discussioncomment-2142511) UR actually extracts 6D parameters of each joint instead of D-H parameters. And use it to directly generate URDF model instead of correcting the default one.
> 
> Which information is currently possible to extract from the robot using MotoSDK? Is it possible to extract information of calibrated kinematics from Absolute Accuracy? Is it possible to extract D-H parameters? If true, where information can be found about its definition for Yaskawa robots (similar to [UR](https://www.universal-robots.com/articles/ur/application-installation/dh-parameters-for-calculations-of-kinematics-and-dynamics/)).

As @ted-miller already mentioned, it is possible to retrieve the D-H parameters for a particular group. See the mentioned `GP_getDhParameters(..)`:

https://github.com/Yaskawa-Global/motoros2/blob/10d8f00317c5d4ef7960627962337bf72e2ba817/src/CmosParameterExtraction.h#L196-L205

At some point during MotoROS2 development, I had an initial implementation of a service which made this information available to the ROS side. That never got merged.

Even though @ted-miller wrote (https://github.com/Yaskawa-Global/motoros2/issues/206#issuecomment-1924771500):

> [..] the DH parameters for the robot never get updated after a calibration. Normally, we only modify the encoder's zero-position when calibrating.

it might still be a good idea to implement that service.

**Comment by iydv** [comment]
Today we have done some investigation regarding D-H parameters from Absolute Accuracy software. Unfortunately, the values obtained after the calibration are stored in different file and cannot be retrieved using this function. 
Either this function has to be updated or we need to use different methods to retrieve updated D-H parameters. 

I think there is a system parameter which indicates that Absolute Accuracy is installed. Probably the service would have to check it and load the data from new file instead of the default one.

**Reply by gavanderhoorn** [reply]
@iydv: could you please describe what you have tried and discovered? It would be valuable to have a log of your efforts here.

**Reply by gavanderhoorn** [reply]
Friendly 'ping' @iydv and @yeu-geissdoerfer?

**Comment by yeu-geissdoerfer** [comment]
I want to first share some **background information** on the possibility to "calibrate URDFs".

If a customer wants to use "Absolute Accuracy Compensation", the robot is surveyed in absolute terms. This measurement results are saved in an ABSPACC.DAT and give information about the *"real" DH parameters* and the robots spring constant. 

This *"real" DH parameters* are used to generate a **"calibrated URDF"** that represents a **more realistic kinematic description of the robot**.


Currently we are testing a *"calibrated URDF"* and share more information later.

**Reply by gavanderhoorn** [reply]
thanks for the update.

> Currently we are testing a "calibrated URDF" and share more information later.

could you already say something about how you created said URDF? Which parts did you modify? Did you use any existing approaches? Have you already thought about how to deal with IK solvers?

Iteratively prototyping things and testing as much as possible while keeping others in-the-loop avoids potentially having to go back and reimplement things later. It'd be great if we could already discuss approaches.

**Comment by yeu-geissdoerfer** [comment]
**Note: This is the procedure used in an alpha test and doesn't state a final procedure!**

The **_current approach_** for creating a "calibrated URDF" is as follows:

1. Calculate the **_DH parameter_** using the ABSPACC.DAT
2. Calculate the **_differences of `alpha` and `theta` values_** with the following formula `URDF alpha = - (alpha from default DH - alpha from corrected DH)`
3. Mapping the **_DH parameters to URDF_**
3.1 `alpha `= `yaw`
3.2 `a `= `z`
3.3 `d` = `x`
3.4 `theta` = `roll`
4. Change default DH parameter to calibrated DH parameter in robot controller (only possible with Yaskawa Mode)

Now the question comes up. Why calculating only the difference of `alpha `and `theta `and not of `a` and `d`?
By looking at a default URDF we can see that `roll, pitch, yaw` are always `0`. So when creating a "calibrated URDF" we only need to account for the rotational difference. `a` and `d` represent the absolute link length. Using only the difference for `a` and `d` would cause the robot to shrink.

It is important to update the DH parameter in the robot controller. Otherwise the robot controller outputs wrong position values. Unfortunally this is only possible with an Yaskawa Mode. There is the possibility that our Hotline Service can create a ALL.PRM file with calibrated DH parameter, that you can then load without Yaskawa Mode.

**Reply by ted-miller** [reply]
> Calculate the DH parameter using the ABSPACC.DAT

Does this also require a Yaskawa technician? Can it be performed by a standard user?

**Reply by gavanderhoorn** [reply]
Friendly ping @yeu-geissdoerfer?

**Reply by yeu-geissdoerfer** [reply]
We will provide a calculation soon

**Reply by gavanderhoorn** [reply]
Friendly ping @yeu-geissdoerfer

**Comment by yeu-geissdoerfer** [comment]
For the **_calculation of the DH parameter_**, I provide an executable here in this comment. _Simply place the ABSPACC.DAT file in the same folder as the executable and you will receive a DHTable.txt as output_. Please do not change the name of ABSPACC.DAT, as the robot controller and the executable expect this name convention.

Note: The **_zip folder contains the executable_**, as an exe file cannot be added to a GitHub comment.
[DetermineDHParameter.zip](https://github.com/Yaskawa-Global/motoros2/files/14959473/DetermineDHParameter.zip)

**Reply by gavanderhoorn** [reply]
Thanks for posting that @yeu-geissdoerfer and thanks for developing the tool.

Could you confirm your description of the process as you gave it in https://github.com/Yaskawa-Global/motoros2/discussions/210#discussioncomment-8643412 is still accurate?

Specifically your comments about *Yaskawa Mode* being required?

**Reply by yeu-geissdoerfer** [reply]
> Note: This is the procedure used in an alpha test and doesn't state a final procedure!
> The current approach for creating a "calibrated URDF" is as follows:

This procedure is still valid. The executable is performing only step 1 of this procedure.

Regarding _Yaskawa Mode_ required, is still valid for step 4. Here the workaround would be that you can call the service hotline, tell them the DH parameter you want to get changed, they create an validated ALL.PRM file, which can then be uploaded without _Yaskawa Mode_.
